feat: allow git remote names in gh repo set-default

majiayu000 opened the pull request:
## Summary
When specifying a repository for `gh repo set-default`, users can now use a git remote name (e.g., "origin", "upstream") instead of the full OWNER/REPO format.

## Changes
- Modified argument parsing to first check if the argument is a git remote name
- If found as a remote name, uses the corresponding repository
- Falls back to OWNER/REPO format if not found as a remote
- Added documentation example for using remote names
- Added test case for remote name argument

## Example
```shell
# Before: had to use full repo format
gh repo set-default myorg/some-repo-name

# Now: can also use remote name
gh repo set-default origin
```

## Test plan
- [x] Added test case for "remote name argument" 
- [x] All existing tests pass
- [x] Updated help example to show remote name usage

Fixes #9149

majiayu000 committed:
feat: allow git remote names in gh repo set-default

When specifying a repository for `gh repo set-default`, users can now
use a git remote name (e.g., "origin", "upstream") instead of the
full OWNER/REPO format.

The command first checks if the argument is a git remote name. If
found, it uses the corresponding repository. Otherwise, it falls
back to parsing the argument as OWNER/REPO format.

Example:
  gh repo set-default origin

Fixes #9149

Signed-off-by: majiayu000 <1835304752@qq.com>

babakks commented on pkg/cmd/repo/setdefault/setdefault.go:
```diff
@@ -79,24 +82,35 @@ func NewCmdSetDefault(f *cmdutil.Factory, runF func(*SetDefaultOptions) error) *
 		`),
 		Args: cobra.MaximumNArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if isLocal, err := opts.GitClient.IsLocalGitRepo(cmd.Context()); err != nil {
+				return err
+			} else if !isLocal {
+				return errors.New("must be run from inside a git repository")
+			}
```
Thanks for pulling this up! 🍻

babakks commented on pkg/cmd/repo/setdefault/setdefault.go:
```diff
@@ -79,24 +82,35 @@ func NewCmdSetDefault(f *cmdutil.Factory, runF func(*SetDefaultOptions) error) *
+					opts.Repo, err = ghrepo.FromFullName(args[0])
+					if err != nil {
+						return err
+					}
```
Please note that git remotes can have `/` in their names. It's probably rare, but still allowed. The current change will not pick them up.

Also, even if we allow `/` in remote names, the current ordering prioritises remotes over repo names, which can be a breaking change in cases where people already have remote names with `/` in them. Again, it's rare, but still valid. Also, semantically looking, checking for `/` in the given arg is kind of leaking the internals of the `ghrepo.FromFullName` into the command level.

So, I'd suggest we try `ghrepo.FromFullName` first, and if we get an error, then try to check if it's a valid remote name. If both checks returned an error we should return a explanatory error like:

```
given arg is not a valid repo or git remote: <error from FromFullName>
```

Also, to confirm this behaviour will not break in future, we should add a special test case where there's a git remote named `OWNER/REPO` that points to a URL different than `gihtub.com/OWNER/REPO`, and expect to see `github.com/OWNER/REPO` is ultimately gets set as the default repo.

majiayu000 replied:
Thanks for the detailed feedback! I updated the argument parsing to try ghrepo.FromFullName first and only fall back to remote name
lookup on error, with the combined error message when both fail. I also added a test case where a remote named OWNER/REPO points
elsewhere to ensure the repo name still wins. Please take another look.

babakks reviewed (changes requested):
Thanks for the PR, @majiayu000! 🙏 It's almost there, except for a small modification.

I haven't yet fully reviewed the tests, though.

majiayu000 committed:
Handle repo argument before remote name

majiayu000 committed:
Merge branch 'trunk' into fix-9149-set-default-remote-name

babakks commented on pkg/cmd/repo/setdefault/setdefault.go:
```diff
@@ -79,24 +82,37 @@ func NewCmdSetDefault(f *cmdutil.Factory, runF func(*SetDefaultOptions) error) *
-					return err
+					if opts.Remotes == nil {
+						return fmt.Errorf("given arg is not a valid repo or git remote: %w", err)
+					}
```
This check is redundant/confusing as `opts.Remotes` is always assigned with a non-nil value here:

https://github.com/cli/cli/blob/c858c524b1777a14e9d89dd50648fcded4227289/pkg/cmd/factory/default.go#L40

babakks reviewed (changes requested):
Thanks for the fix, @majiayu000! 🍻

We just need to get rid of the redundant `if`-statement (see the first/only comment), and then I think we're good to ship this.

---

Just as a bit of context, most `gh` commands work in a two-stage kind of way: 1) pre-run (e.g. arg/flag validation, which are done within `cmd.RunE`), and 2) run (done in a top-level `runX` function). In most cases, we restrict the pre-run stage to only static/offline checks that do not need any API calls, or git invocations.

The `repo set-default` command has already been a bit of an exception due to the call to `opts.GitClient.IsLocalGitRepo` which invokes `git rev-parse` under the hood. Now with the new remote lookup step (in `cmd.RunE`) we're bending the rules a bit more.

That said, I'm not going to ask you for the improvement in this PR, of course unless you're up for it, in which case please let me know.

Anyway, I'm going to capture what should be done here:
1. In pre-run, try parsing the arg as a repo name.
   - If it's not, just leave `opts.Repo` unassigned (i.e. `nil`).
2. In run:
   1. Check current directory is a git (sub)dir.
      - If it's not, bail out.
   2. if `opts.Repo` is `nil`, try parsing the arg as a remote.
      - if it's not, bail out with an error like "given arg is not a valid repo or git remote".

This change will require updating the tests, but has the merit of making the pre-run stage purely static/offline.

majiayu000 committed:
fix: simplify set-default remote parsing

majiayu000 commented:
Fixed the redundant Remotes nil check (factory always sets it) and adjusted tests to always provide a Remotes func. Pushed to fix-9149-set-default-remote-name.

majiayu000 reviewed (commented):
Addressed: removed redundant Remotes nil check; tests now always supply Remotes func. Ready for re-review.

babakks reviewed (approved):
LGTM! Thanks, @majiayu000! 🍻

RodnyPHP reviewed (commented):
-
